misc: Fix error handling

dalcinl opened the pull request:
## Pull Request Description

### attr: Fix cleanup on errors from user delete callbacks

* Prevent memory leaks upon failures in attribute list duplication.

### errhan: Fix user callback error pass-through

Handle consistently the return error code from user callbacks in
copy/delete attributes and query/cancel/free generalized requests.

### errhan: Better handling of user callback errors

* If a user callback returns an non-MPI error code, put it in the error
  string but do not return it back to the caller of the outer MPI routine.
  Otherwise, inspecting a non-MPI error code for error class or string
  is doomed to fail.

* If a user callback returns a predefined error class, still create and
  return error code of that class. This way, the error string will contain
  the stack trace, and the caller can still get the error class out of
  the returned error code.

* Dynamically defined error classes/codes pass through from callbacks to
  the caller of the outer MPI routine. Unfortunately, inspecting the error
  string will not provide the stack trace.

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [ ] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

dalcinl commented:
mpi4py tests are good, [running to double check](https://github.com/mpi4py/mpi4py-testing/actions/runs/8169815927) with added test from https://github.com/mpi4py/mpi4py/commit/e5bf77fb080b2e345dba6fd8497910b2535e2581

PS: I have not run MPICH test.

hzhou commented:
test:mpich/ch3/tcp
test:mpich/ch4/ofi

dalcinl commented:
In one of the failing MPICH tests there is the following comment:
```c
    /* Because the attribute delete function should fail, the attribute
     * should *not be removed* */
```

I'm not sure if you look carefully to my comments in the commit messages. The reason for the failure is that I'm indeed changing the behavior described in the comment above. If a delete function fails, what else a user or MPI can do to recover from that situation? Trying to remove the attribute again may (most likely?) just fail again. Therefore my take was to get rid of the attribute from the list, as that's basically the caller's intention. As long as the error is not silenced (it is not) we are good.

One thing I just realize I did not consider is what to do when setting a new attribute value if the attribute was already set and delete fails. 

Setting a new value needs to invoke the delete callback first on the old value, then what to do if delete fails? There are many possible options: 
1. Ignore the delete error and set the new value. Issue: errors are silenced.
2. Set the new value and error. Issue: confusing, the failure was not in setting new but in deleting old
3. Keep the old value and error.  Issue: similar to trying to delete a value. The old value is bogus.
4. Remove the old value, not set the new, and error. I believe this is the behavior from the changes in this PR, but I'm not 100% sure yet. It is admittedly awkward behavior, I don't like it.

With the purpose of inspiration, let me add a comment about how Python handles error on cleanups. When the reference count of an object drops to zero and the object has to be deallocated, the `__del__` destructor can raise an exception. In such cases, Python keeps going with the deallocation and gets rid of the object for good, but prints to stderr,  [docs here](https://docs.python.org/3/c-api/exceptions.html#c.PyErr_WriteUnraisable). 
Maybe this is the proper way to handle delete attr errors when setting a new attr value? This would be basically option 1 in the list above, except that the error is not fully silenced and a messages goes to stderr.

dalcinl commented:
After a read of the MPI standard, I note the following:

> If an attribute copy function or attribute delete function returns other than
MPI_SUCCESS, then the call that caused it to be invoked (for example, MPI_COMM_FREE)
is erroneous.

Therefore, if a `MPI_Xxx_set_attr` call invokes an attribute delete_fn function that fails, then `MPI_XXX_set_attr` must fail, and similarly for any related MPI class. What the standard does not say is what the state of the attribute list is after the `MPI_Xxx_set_attr` call, so we have to consider that implementation dependent. In this PR I'm changing behavior such that old values are discarded and the new value is always set. I'm updating the MPI test suite to be able to cope with both behaviors.

hzhou commented:
> Setting a new value needs to invoke the delete callback first on the old value, then what to do if delete fails?

If the delete callback return failure, it means --
1. User made a mistake in their callback function; user should fix their program!
2. User intend to prevent the old attribute from being removed. Then `MPI_Xxx_set_attr` should fail. This can be a valid use case where User dynamically decide when an attribute can be safely removed -- essentially an `EAGAIN` error.

dalcinl commented:
> 1. User made a mistake in their callback function; user should fix their program!

Errors are not not necessarily the user's fault. Errors may be the consequence or runtime conditions. In my particular use case, mpi4py is acting as middle man trampolining from user code in Python to MPICH code in C. Even if errors are the consequence of end users mistakes, the MPI library developer (or Python/Julia/Java language binding developer) must do its best to make these error non-fatal, and allow for recovery leavion the internal MPI library state consistent.

> 2. User intend to prevent the old attribute from being removed. Then `MPI_Xxx_set_attr` should fail. This can be a valid use case where User dynamically decide when an attribute can be safely removed -- essentially an `EAGAIN` error.

I cannot buy this argument. Making a delete/destructor routine fail to flag an attribute as readonly is a total abuse of semantics. Do you have any real life example of such a use case? Preventing explicit attribute removal with delattr/setatr, if there are compelling use cases, should be added as a new feature in the upcoming MPI standard. Otherwise, conflating a destructor with readonly flagging sounds simply as vile hack, and supporting such a thing would interfere with what IMHO are a more compelling (and already existing) use case I'll describe in the following .

Look, the reason I'm making all this points and put this PR together is to fix outstanding issues in mpi4py related to attribute management. When user callbacks fails, I have to perform cleanup at the Python level (typically, this mean decrement the reference count of some Python objects). Allowing for a delete function to be called twice put me in an impossible situation in case of failures. If there is an error, should I cleanup things or not? If I perform cleanup, and the delete function is ever called again, then the internal state is not valid. If I do not perform cleanup and the delete function is not called again (or fails again), then I leak resources.  Unfortunately, a delete callback does not know if it is being invoked because of delattr, setattr, or container object being freed, therefore I cannot now what's the proper way to handle errors in each case. Of course, I could just catch errors, print to stdout, and return success from the callback. But that is breaking the rule that errors should not pass silently. A warning in stdout is not completely silent, but has no match for an error notified the proper way. Please note that similar issues would apply to any other language binding that has to manage language-specific state associated with the attribute value, it is not just about my silly toy Python thing.

A few days back, I was having a talk with @jeffhammond about how the C MPI API/ABI should evolve to allow for other languages (including Fortran) to be able to wrap MPI in a clean way. All the MPI parts that involve user-defined state and callbacks have to be revisited, with the exception of generalized request that already does the right thing. Within  generalized requests, the standard is quite clear that the free_fn callback is invoked only once. Makes total sense! The free_fn is a destructor routine for a user-defined context! In the case of attributes, a similar logic should be followed. An attribute delete callback is a destructor for the attribute value, destruction should be its only purpose, and it should be called only once by the MPI library. A consequence of this rule is that if an attribute delete callback fails, you have to consider the attribute value irremediably lost, remove any reference to it form the attribute list, and report the error back to the caller.

@hzhou I'll wait for your new comments with a rebuttal to my arguments above. If you insist on keeping the old semantics, I'll stop fighting you and update this PR accordingly. Otherwise, the previous test issues are now resolved and this PR should be ready to go.

hzhou commented:
>> 1.  User made a mistake in their callback function; user should fix their program!
>
> Errors are not not necessarily the user's fault.
> 
To avoid confusion, we are talking about a user-defined callback function that returns a non-`MPI_SUCCESS` code, right? Since user defines the function as such, we need assume the user has a valid reason for it. It could be a *coding* error, in which case, it should be corrected. Or it could be a runtime situation, then user code need to resolve this runtime situation. MPI library has no knowledge of such runtime situation and cannot do anything. IMO, it is wrong to *ignore* the user-defined return code. I don't have an actual use case, but I can imagine a user program could create an object and attached to a `comm` as an attribute; the object can be reference counted to track its usage or dependency; thus it can refuse to be deleted if the reference count is not cleared. The reason for it be a runtime dependency or ordering issue that only user can resolve. But if MPI library ignore it and always blindly remove the attribute, it defeats the user's design.

Callbacks are tricky regarding language bindings, especially when the language adds extra semantics that MPI did not account for. For example, there are subtle differences in how we treat C callbacks and Fortran callbacks. Maybe we need some special treatment for the Python bindings as well?

To be honest, I am not exactly understand the Python situation. Can you give me a specific example? Is the callback (attr delete function) defined by end users (in Python routine) or is it a wrapper function defined in the MPI4Py implementation?

hzhou commented:
Thanks for the explanation, I think I understand the situation better now. The reply is getting longer. Maybe we should have an offline virtual meeting? But here is my current view --

If the user callback (in Python) returns a non-SUCCESS code, then that code should be returned in MPI4Py's trampoline callback, and we should not delete the attribute since that is the end user's intention.

If there is an error in MPI4Py or a runtime error handled by MPI4Py, then MPI4Py's trampoline callback should return MPI_SUCCESS because, to the MPI library, the error has been handled outside MPI and MPI should proceed to delete the attribute. ow mpi4py feedback the error to the user is mpi4py's decision -- I agree it shouldn't be a silent error.

hzhou commented:
> If you agree with me that a artribute delete function is basically a destructor, then you will have to agree with me that destructors cannot be invoked twice on the same object. Transitively, if a destructor fails, then you have to move own, forget about it, remove the thing from any container, and REPORT BACK THE ERROR.

It can serve as a destructor. That's one of usage. But it can be more flexible than a destructor. A destructor cannot return a code. An MPI callback is designed to have a return code and thus can serve more complex runtime situations.

dalcinl commented:
> A destructor cannot return a code ... If you intend the delete function to serve the sole purpose of a destructor, then it should always return MPI_SUCCESS.

Note however that the free callback of generalized requests do return a code, and AFAICT its intention it to act as a destructor.

> The reply is getting longer. Maybe we should have an offline virtual meeting?

OK, good idea. We should involve Jeff, too. IMHO this issue is not about Python, but rather a general API misfeature.

In the meantime, I'm thinking on splitting this PR in two to get the non-controversial parts of it merged in first.

dalcinl commented on src/mpi/attr/attrutil.c:
```diff
@@ -216,14 +208,19 @@ int MPIR_Attr_dup_list(int handle, MPIR_Attribute * old_attrs, MPIR_Attribute **
   fn_exit:
     return mpi_errno;
   fn_fail:
+    /* In case of any failure, remove the attributes duplicated so far
```
This one is important. If during the duplication of an attribute list, any of the user-define copy_fn functions fail, then I attempt to deallocate the list duplicated so far. Note however that if any delete_attr_fn function fails, I just ignore that error, IMHO reporting the one from copy_fn should take precedence.

dalcinl commented on src/mpi/attr/attrutil.c:
```diff
@@ -242,10 +239,12 @@ int MPIR_Attr_delete_list(int handle, MPIR_Attribute ** attr)
+         * corresponding keyval. If the delete function fails,
+         * we record the last failure */
+         rc = MPIR_Call_attr_delete(handle, p);
+         if (rc != 0) {
```
This this trivial change we can report the first failure.

```suggestion
         if (mpi_errno == MPI_SUCCESS && rc != 0) {
```

hzhou replied:
The suggestion sounds okay

dalcinl commented:
The last force-push was just to fix an outdated commit message.

hzhou commented:
test:mpich/ch3/tcp
test:mpich/ch4/ofi

hzhou commented on test/mpi/attr/attrerr.c:
```diff
@@ -99,14 +99,34 @@ int test_communicators(void)
+            printf("Error with second put\n");
+        }
+    }
+
```
Quoting the text from MPI-4.1 specification:
> This function (`comm_delete_attr_fn`) is called by MPI_COMM_FREE, MPI_COMM_DISCONNECT, MPI_COMM_DELETE_ATTR, and MPI_COMM_SET_ATTR to do whatever is needed to remove an attribute.
> The function returns MPI_SUCCESS on success and an error code on failure (in which case MPI_COMM_FREE will fail).
>
I am not sure why it only mentions `MPI_COMM_FREE` will fail, but I believe it supports that `MPI_COMM_DELETE_ATTR` and `MPI_COMM_SET_ATTR` also should fail. By fail, I think it commonly will be interpreted as the operation of deleting the attribute is not successful, i.e. not removed.

However, I concede that the exact behavior under an failure case is not well defined. So I guess your interpretation here is acceptable.

dalcinl replied:
@hzhou Lucky those that mostly care about the latest standard. Take the time machine back to MPI-1.1, and you will see the exact same sentence. Therefore, this text only mentioning Comm (and not Datatype nor Win) is a remnant of the time only communicators had attributes.

dalcinl replied:
> However, I concede that the exact behavior under an failure case is not well defined.

And that's the root of the problem. If you leave things open to interpretation, implementations will vary wildly. I just remember a long time ago discussion we had elsewhere with other people about how to interpret lower/upper bounds of datatypes and its propagation to new datatypes.

hzhou replied:
Undefined is not the same as leaving to open interpretation. Often it is undefined because it is not possible or sensible to define one way or another. Take the error case, the failure may prevent us from removing the attribute. Yeah, we could blindly remove the attribute by ignoring the error return, but if the error return is due to some runtime exception, it may mean users didn't finish their cleanup thus we are creating a memory leak by removing the reference (attribute). But it is possible that the failure does not affect the cleanup or that there is nothing there is to be done for the cleanup. So undefined only have one clear message -- try to *avoid*. :)

dalcinl replied:
I'm not sure whether you are acknowledging that the MPI spec is vague/deficient/incomplete, or you are defending it as it is now. In any case, we will never reach an agreement in this thing, we have "axiomatic" differences that are irreconcilable.

dalcinl commented on src/mpi/attr/attr_impl.c:
```diff
@@ -594,15 +597,15 @@ int MPIR_Type_set_attr_impl(MPIR_Datatype * type_ptr, MPII_Keyval * keyval_ptr,
+            p->value = (MPII_Attr_val_t) (intptr_t) attribute_val;
+            /* Does not change the reference count on the keyval */
             break;
         } else if (p->keyval->handle > keyval_ptr->handle) {
```
@hzhou Can you please check this part of the code for datatype (and similarly for windows)? I cannot make sense why this branch is needed, and why things are not done the same way for communicators.

It looks like it is somehow related to wanting to insert in a keyval order in the list rather than in the tail . I have no objections to that, really, but I found the asymmetry in communicators striking. I noticed the comment /* ``/* CHANGE FOR MPI 2.2: If not found, add at the beginning */``, but I don't remember any specific standard mandate about that.

hzhou replied:
All the attributes are linearly ordered in linked list. This branch is where we found the position where the new entry need be inserted. Now that I am looking at this code, seems there is a bug that we are inserting after `p` but seems we should insert before `p`. 

Apparently, the comm path was changed at some point and we neglected to apply the changes to Type and Win. I don't know the original reason for the change. I'll follow up.

hzhou replied:
Found the original commit:
```
commit 8f959cd7bf79dd335a0d5d439643bafa4080677e (HEAD)
Author: David Goodell <goodell@mcs.anl.gov>
Date:   Thu Aug 20 20:45:04 2009 +0000

    [svn-r5250] Attribute delete callbacks are now ordered (MPI-2.2 ticket #71).

    Implementation of MPI-2.2 ticket #71.  Delete callbacks are now invoked
    in LIFO order.  Once the MPI version number macros are updated to 2.2
    the test added in r4346 will check for this behavior.  This change also
    includes another test for LIFO ordering because I wrote it before I saw
    Bill's original test.

    For more information and history on this issue, please see this URL:

    https://svn.mpi-forum.org/trac/mpi-forum-web/ticket/71

    The non-testing portions of this change were originally written by Bill Gropp.

 src/mpi/attr/comm_set_attr.c |  24 +++++---------------
 test/mpi/attr/Makefile.sm    |   1 +
 test/mpi/attr/attrend2.c     | 128 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 test/mpi/attr/testlist       |   2 ++
 4 files changed, 137 insertions(+), 18 deletions(-)
```

hzhou replied:
The comment in the test is clearer:
```
/*
      The MPI-2.2 specification makes it clear that attributes are called on
      MPI_COMM_WORLD and MPI_COMM_SELF at the very beginning of MPI_Finalize in
      LIFO order with respect to the order in which they are set.  This is
      useful for tools that want to perform the MPI equivalent of an "at_exit"
      action.
 */
```
I guess that is the reason the change was only applied to the comm, but not type or win. I think we should update the type and win as well to make it consistent.

dalcinl replied:
That comment is a inaccurate, though. The MPI 2.2 standard only talks about such ordering for `MPI_COMM_SELF` only, and that's still the case in MPI-4.1.
Anyway, I think your intention of making everything consistent makes sense, despite the small 0.5 performance penalty on the average for objects with lots and lots of attributes (with should be an extremely rare case).
Alternatively, you could just special case `MPI_COMM_SELF` and skip the branch to force the LIFO order only for that case. I think I still prefer LIFO everywhere, just my two cents.

dalcinl commented:
@hzhou Should we rebase and test again?

hzhou commented:
> @hzhou Should we rebase and test again?

It is not necessary. The tests are clean. We can rebase it before merging. I'll finish the review soon.

hzhou commented on src/mpi/errhan/errutil.c:
```diff
@@ -1024,16 +1024,21 @@ int MPIR_Err_create_code_valist(int lastcode, int fatal, const char fcname[],
+    } else if (is_valid_error_class(lastcode) || (lastcode & ERROR_DYN_MASK)) {
+        err_code = lastcode;
+    } else {
+        err_code = error_class;
```
Could you explain this change? (maybe we should schedule a call to walk me through).

dalcinl replied:
The changes here are to mimic the changes in the other conditionally-compiled routine with the same name. We want the `lastcode` from user callbacks to pass through, but only if its value corresponds to a valid error class, otherwise jut return `error_class`.

dalcinl replied:
In other words, the last two branches are just an attempt to NOT pass through values that are unknown to MPI. To pass through, `lastcode` has to either b A) a predefined error class `MPI_ERR_*`, or B) a dynamically created error class or code. If a user callback returns let say `return 123456`, that thing will not pass through, and MPI will eventually report back the `error_class` that was passed to  `MPIR_Err_create_code_valist()`.

dalcinl commented:
@hzhou Gentle reminder to finish reviewing this PR. Were my previous explanations good enough?

hzhou commented on src/mpi/errhan/errutil.c:
```diff
@@ -1024,16 +1024,21 @@ int MPIR_Err_create_code_valist(int lastcode, int fatal, const char fcname[],
-            errcode = (generic_idx << ERROR_GENERIC_SHIFT) | error_class;
-            if (fatal)
-                errcode |= ERROR_FATAL_MASK;
+            err_code += (generic_idx + 1) << ERROR_GENERIC_SHIFT;
```
Fixing a bug that we need generic_idx + 1

hzhou replied:
@dalcinl The bug fix need be in separate commit; white space change in separate commit; spelling changes in separate commit; macro refactors in separate commit; actual changes in separate commit. I understand this may seem a hassle for contributors, but it helps the review *tremendously*, and ultimately avoids wasting contributor's time due to review delays.

dalcinl replied:
Don't forget that the ultimate time-saver for contributors is to not contribute. Unfortunately, next time I'll consider contribution I'll have to weigh in the benefits of providing cleanups and consolidations rather than just focusing in the minimal set of changes to get the fix I need accepted with no objections. 

This time, I consider your request to further split commits is unnecessary burden on me. The base, starting code I  had to deal with was old, there were bugs, had to review more modern code to consolidate style, come up with proper fixes while doing exhaustive testing. The whole thing was painful and unrewarding. And then I'm asked to roll back git history and go trough additional pain to further split commits. Such requests are reasonable in general, but in this particular instance, I honestly think it is unreasonable. Nonetheless, I will try do it. By engraving the pain in my brain, and I will learn my lesson.

hzhou commented on src/mpi/attr/attrutil.c:
```diff
@@ -97,11 +96,8 @@ int MPIR_Call_attr_delete(int handle, MPIR_Attribute * attr_p)
-            MPIR_Err_create_code(MPI_SUCCESS, MPIR_ERR_RECOVERABLE, __func__, __LINE__,
-                                 MPI_ERR_OTHER, "**user", "**userdel %d", rc);
 #endif
+        MPIR_ERR_SET1(mpi_errno, MPI_ERR_OTHER, "**user", "**userdel %d", rc);
```
Looks like you always want to set a new error code here. I thought you wanted to pass through the user `rc`. What did I miss?

dalcinl replied:
The line right above the removed lines sets `mpi_errno = rc`. And then the next line is `MPIR_ERR_SET1(mpi_errno, ...`. Therefore the error do pass trough.

hzhou replied:
`MPIR_ERR_SET1` creates a new error code, the old error code is stored in the error ring so we can show error stack on abort, but users won't be able to access that.

dalcinl replied:
I'd prefer to wait for #6950 to be merged to update this one and keep discussing about it. I tested this PR extensively with mpi4py, error propagation was definitely working fine. Unfortunately, all that testing is no longer relevant, as I had to completely change the way mpi4py handles errors in callbacks to not rely on MPI passing errors through.

dalcinl replied:
> `MPIR_ERR_SET1` creates a new error code, the old error code is stored in the error ring so we can show error stack on abort, but users won't be able to access that.

Sorry, Hui, it is complicated, but to my defense, this is not my code. You are not seeing the full picture of the error handling implementation in MPICH. The `MPIR_*` routines return some error code that is generated by using the various error macros like `MPIR_ERR_SET1`. You are right, this macro creates a new error code. The user error code get's eventually stored in the error ring and printed to the string, but the error ring definitely preserves the value. In the `MPI_*` layer, if the `MPIR_*` layer returns with error, then MPI error handling takes over, and eventually, the user error code is extracted from the ring and that's what the `MPI_*` caller gets.

It took me quite a while to figure out how all this machinery worked, fixing and consolidating stuff in a way that is easy to review is really hard, and that's the reason I imploring you to relax your rules for contribution in this particular instance. Anyway, I'm working right now on splitting one of my commits. Let's see how it goes.

hzhou replied:
> I imploring you to relax your rules for contribution in this particular instance. Anyway, I'm working right now on splitting one of my commits. Let's see how it goes.

It's not a strict rule, but a guideline. I literally not getting the points of your fix and cannot finish the review or even to ask the right question, partly due to the commits are not clean. Splitting the commits allow reviewer to ignore the non-related changes and zoom in the the key contribution of the PR.

dalcinl replied:
> partly due to the commits are not clean. 

After building on top of your commit, I did some refactoring of my commits. Now there is one commit exclusively dedicated to to style/comments/whitespaces that do not change behavior at all, but just tries to consolidate things and match code in the implementations of comm/type/win. The really important commits are 82f5d69977e9b119c28f62baaead7ac677b8514b and  87a3f4750a397bce19225863449208a91ee02223, all the other are mostly preparatory cleanups.  Commit 4b9959ff2b6f8600e641b01bce0861e1809c9bcc  is admittedly involved, but it is what it takes to implement what the commit message describes.

All these commits are IMHO very clean and self contained. You may still not understand them, but it is really because of faulty style on my side, or just because the thing is really complicated? At this point I've spent hours looking at how all the pieces glue together, and there is no way you can grasp what's going on by just looking at the diff. How are we going to get out of this stagnation point?

Perhaps we should open a new PR with the these commits to start fresh and discard all the mess of intermixed comments, at this point it is very confusing. It would also be very constructive if you take a look at my comments below https://github.com/pmodels/mpich/pull/6935#issuecomment-2022721045 and tell me if you agree at least with the spirit of my intentions, otherwise there is no point in moving forward. Then maybe we can negotiate a more complete test case that exercises the behavior I described in full, such that a debugger can walk you through the mess and you can ultimately convince yourself that the implementation is sound.

hzhou reviewed (commented):
Reviewed `attrutil.c`. See my question.

hzhou commented:
Xref: https://github.com/pmodels/mpich/pull/6950

dalcinl commented:
I rebased this PR on top of your commit from #6950. I believe all MPICH tests should now pass. All mpi4py tests pass.

The exact pass-trough behavior I intended is the following:

* If a callback returns a *predefined error class*, that is, any of the `MPI_ERR_*` values, then the `MPI_*` layer routine eventually return an *error code* of the *same class* as the `MPI_ERR_*` class returned by the callback. From the *error code* you can query for the error string to get the usual nicely populated stack trace. Why doing it this way? Well, returning a new error code (of the same class) rather than the error class returned by the callback allows for getting the detailed error string with populated stack trace, and the error class can ultimately be recovered with `MPI_Error_class`.
* If instead the callback returns a *dynamic* error class or code, then that error code/class is passed through unmodified. The `MPI_*` layer will get the dynamic error class/code, and querying for the string will return whatever string the user has registered for that error class/code. Why this behavior? Well, if the user goes through the hassle of creating an error class/code to return it form a callback, then I'd rather not second-guess the user and return the same exact thing without playing games. AFAICT, I'm not sure MPICH has any test for this behavior. mpi4py certainly does not have a test for it. 
* If the callback return some nonzero garbage integer value that cannot be interpreted as either a builtin error class or a dynamic error class/code, then the return value will be printed in the error string, but the concrete returned error from the `MPI_*` layer will come from elsewhere (built from the `error_class` argument). Why this behavior? Well, if the return is garbage, letting that value pass trough and escape the `MPI_*` layer may trigger a different error at a later point if the users tries to for example query for an error string.

I've put a lot of effort to untangle the mess and come up with the semantics I proposed above with the intention providing MPICH users with some more featured pass-trough behavior for user callback errors. Now, I totally understand the core MPICH development team may not like or not care about all what I wrote above and the code changes I implemented in this PR. In such a case this PR can be closed, I'm not going to keep fighting about semantics for a software project I'm not even a member of, you guys definitely know better what's best for MPICH and its user base. Of course, if anyone believes that the actual code implementation does not really match what I described above, then the next step is to prove me wrong with an actual reproducer, and I'll happily acknowledge my mistake and will proceed to provide a fix as soon as possible.

dalcinl commented:
We are going nowhere. I'll wait until you finish #6950 and then I'll see what what can I do.

hzhou commented:
@dalcinl Sorry for lack of response. I am working on #6950 to understand the code and hoping it can help me to understand your code and have a more informed discussion.
